### DISCUSSION TITLE
Custom Musics

### DISCUSSION THREAD

**Discussion by charlyb01** [sf-inc/music_control#17] [Show and tell]
# Custom Musics

As you may know, you can now add your own musics via a Resource Pack! You just have to follow a few steps so that **_Music Control_** will recognize your musics properly. You can find an example down below 😃 

## Custom categories

For each resource pack, custom categories can be created with the name of your resource packs' namespaces.
Musics that are in the `assets\soundpack_namespace\sounds\music` folder will be added to the new category **Soundpack Namespace**. Note that musics shouldn't be placed anywhere else or they won't be detected properly.

Example:
- `assets\soundpack_namespace_1\sounds\music` folder will add the category **Soundpack Namespace 1**
- `assets\soundpack_namespace_2\sounds\music` folder will add the category **Soundpack Namespace 2**
- `assets\soundpack_namespace_x\sounds\records` folder will add music in it to the **Disc** category _(it won't create new discs!)_

## Musics' names

For a nicer display of the custom musics when a new music is being played, you can add them a custom name too! I recommend you to display the name of the music's author and the music's name if you plan to share this soundpack. Indeed, that's a nice and easy move that let people using your soundpack know who and what they are listening to! 😉 
For this, you just have to complete the lang file located in the `assets\soundpack_name\lang` folder as follow (you have to use the filenames without the `.ogg` extension). Note that you don't have to create several lang files as the default one will be used in every language. However, you should create one in each namespace (not like in the following example)
```json
{
    "soundpack_namespace_1:music/my_custom_music1": "Author1 - My Custom Music 1",
    "soundpack_namespace_2:music/my_custom_music2": "Author2 - My Custom Music 2"
    "soundpack_namespace_x:records/my_custom_music3": "Author3 - My Custom Music 3"
}
```

## Music not working?

Here is a non-exhaustive list of common mistakes.

* **`sounds.json` not correct:** Json files rely on proper `{`, `}`, `[`, `]` use.  
  Make sure every opening braces have their closing braces.  
  I would recommend using a nice text editor / IDE to make your life is easier, such as Notepad++ or Visual Studio Code.
* **Category not correct:** Do not change the `"category" : "music"` in your `sounds.json`.  
  This category is used by the data packs to tell in which channel it should be played *(see Minecraft "Music & Sounds" options)*.
* **Filepaths not recognized:** Minecraft can only read `[a-z0-9/._-]` character in path.  
  Avoid using special characters, this advice is valid not only for Minecraft. If wou want to display nice titles, rely on the `lang` file instead.
* **Files hierarchy:** Music files must be in `assets\soundpack_namespace\sounds\music`. Do not forget the last `music` folder.
* **Music file format:** Music must be in `.ogg` format.
* **Music file compression:** The song bitrate must not be too high (don't know the maximum).  
  Some users had issues because they downloaded a music in ogg which was way too big and not recognized by Minecraft.  
  Default Audacity ogg export settings works fine if you think you have this issue. However, if you got all of your songs the same way,
  it's likely not your issue.

I hope you enjoy this new feature!
Feel free to comment about it there, and share your best soundpack!

**Comment by charlyb01** [comment]
[Music Control - More Musics](https://www.curseforge.com/minecraft/texture-packs/music-control-more-musics)
Official mod's Resource Pack! Refer to this one to create one on your own 😉

**Reply by StarzPlatoon** [reply]
Hey, so I am tryna make a pokemon platinum playlist for this map and I have done so much research but cannot seem to figure out how to set this up WHATSOEVER. Is there any ways you could help?

**Comment by charlyb01** [comment]
**_Backlog for Music Control v1.5 (and earlier):_**

## Worlds categories

You can add musics you want to a specific category already included with **_Music Control_**. That means you can add custom musics to **OVERWOLRD**, **NETHER**, **END**, or **DISC** categories.
Musics that are in the: 
* `assets/soundpack_name/sounds/music/game` folder will be added to **OVERWORLD**
* `assets/soundpack_name/sounds/music/nether` folder will be added to **NETHER**
* `assets/soundpack_name/sounds/music/end` folder will be added to **END**
* `assets/soundpack_name/sounds/records` folder will be added to **DISC**

## Custom category

For each resource pack, a custom category will be also created with the name of your resource pack.
Musics that are in the `assets\soundpack_name\sounds\music` folder will be added to this new category.

**Reply by ShuitOnDiscord** [reply]
i was attempting to put all of the "the immersive music mod" songs into a resource pack that music control could use and control, but when attempting to set it up to put specific songs into specific biome categories, such as ocean, swamp, badlands, and so on, it didn't work. i thought maybe this only support the overworld, nether, end, and disc categories, but that also didn't work for me. i'm somewhat confident i am doing something wrong, but the example pack doesn't show how to do what i want so i'm stuck in the dark

**Reply by charlyb01** [reply]
Hi, sorry for the delay, I thought I already replied to you! Doing what you want shouldn't be too difficult. First, you should refer to the proper message: the one you replied to is the old way for v1.5 and older. Then, putting songs into specific events is **not** made into the example resource pack, because this is just an example to **add** songs. To set the events, you should do that through the GUI. You could also that manually if you want, but it's easier to do mistakes

**Reply by ShuitOnDiscord** [reply]
> Hi, sorry for the delay, I thought I already replied to you! Doing what you want shouldn't be too difficult. First, you should refer to the proper message: the one you replied to is the old way for v1.5 and older. Then, putting songs into specific events is **not** made into the example resource pack, because this is just an example to **add** songs. To set the events, you should do that through the GUI. You could also that manually if you want, but it's easier to do mistakes

So there isn't really a way to automatically set it into categories?

**Comment by SmarterCob** [comment]
Question, where would the custom music be located after it is added? I can't seem to find it.

**Reply by charlyb01** [reply]
What do you mean exactly? If you created a Resource Pack as explained above, once you loaded your resource pack in your game, you need to open the music menu to add your music to events, or you can play them directly in the menu. You can search your added music with the search bar.

**Comment by musondanolan** [comment]
Is there a way to disable the name of the song automatically popping up every time it starts playing? I like how most of the soundtracks fade in slowly, so you barely notice they've started.

**Reply by charlyb01** [reply]
Hi,

It's a config of the mod. You can edit them if you have the Mod menu installed, or directly in `.minecraft/config/music_control/cosmetics.json` 😉
It is the `atMusicStart` field that you need to set to `false`.

**Reply by musondanolan** [reply]
Thanks! This is a really great mod!

**Comment by genericbeastt10** [comment]
Hi! Sorry to bother, i'm currently having an issue with making the resourcepack with custom music. I installed the one for reference you added here, added my music as .ogg files, no special characters, no upper case, proper names; Modified the files for the songs.json, and the lang file, but it doesn't seem to work. I left the more_c418 folder and it works as intended, but mine doesn't! I've checked here and there for any issues with the coding on the .json files, or if I screwed up something in the names... it should be working, but i have no idea why it just won't. Is there any help, or advice, you could give me? Thanks in advance!! (I understood that you had to modify the resourcepack manually, please correct me if i'm wrong).
Please, I'm going mad at trying to fix it myself :(

**Reply by charlyb01** [reply]
Hi, do not hesitate to reach me on my Discord if you want more help and share your resource pack 😉 You don't have to keep the `more_c418` folder, but you can if you want. If you add other music, you can add them in another folder, to separate them into music categories. Make sure to keep the same file hierarchy. Each folder you make should have its own `sounds.json` file. In this file, make sure to name the sounds properly, starting with `the_folder_you_made:music/the_song_you_want_without_the_ogg_at_the_end`

**Reply by genericbeastt10** [reply]
Thank you so much!! You were right, I checked, fixed a little error on the sounds.json and lang's file now if I search for it correctly, I'm able to find the music on the menu!!! Thanks a lot :)
(Dumb mistake, it was a coma that didn't allowed the proper code to end!!)

**Comment by ratnm** [comment]
# Make Adding Custom Music Easier

I've made a Python script to simplify adding custom musics with the Music Control mod. Hope this can help you!
The passage below are the environment and usage steps for my script.

**Environment**: Python 3, FFmpeg (ensure it is in your system PATH)

**Minecraft Requirements**: Fabric, Music Control mod, Mod Menu

## Steps

1. Create a new folder as the root directory (this step is important).

2. Inside the root directory, create a `songs` folder and download your music files into it.

3. Place [generate.py](https://github.com/user-attachments/files/26478793/generate.py)  into the root directory and run it.

4. In the program, enter `r` to format the song names (the final names will be displayed as titles in the interface). You can manually rename the songs or modify the renaming logic in the `rename` function inside `generate.py`. When running the formatting function, the program will ask you to confirm the name changes - enter `y` to proceed.

5. Enter `c` in the program to convert the songs to the required format (Minecraft requires `.ogg` files with Vorbis encoding at around 128kbps). The conversion may output a lot of text - this is normal. Do not close the program until it says `done`.

6. Enter `g` in the program to generate the `custom_music` namespace, which will be used later. This will create a directory named `custom_music` containing language files (you can also add your language in the `gen_resourcepack` function inside `generate.py`), `sounds.json`, and all the converted songs.

7. Launch Minecraft, enter any world, press `M` to open the music menu, switch to the **Configure** tab, click **Create a new one**, then click **Save**.

8. In the resource pack selection screen, click **Open Pack Folder**. In File Explorer, navigate to the `music_control/assets` folder, copy the `custom_music` folder generated in your root directory into it, then press `F3 + T` in-game to reload resource packs.

9. Return to the **Configure** tab and turn on the **Music/Event** slider.

10. Find and click the **Dimension: The Overworld** button, then click all the buttons that appear on the right to remove all vanilla music. Turn on the first slider (from left to right) above the text box panel, enter `@custom` in the text box to specify the `custom_music` namespace, and click all the buttons that appear on the right again to add your custom music.

11. Repeat step 10 for the **Dimension: The End** button on the left as well. Finally, click **Save**.

12. Wait a moment, and your game will start playing your custom music, showing the song name above the item bar. If you encounter any problems, feel free to ask me.
